### DISCUSSION TITLE
New Idea - PolyArg Extension Method !!

### DISCUSSION THREAD

**Discussion by sangeethnandakumar** [dotnet/csharplang#7615] [Polls]
# Extending 'Normal Extensions' with 'PolyArg Extensions'

In C# we've normal extension methods that can accept one parameter along the "this" keyword

## Existing Extension Support

```csharp
public static class NormalExtensions
{
    public static bool IsAGoodBoy(this string name)
    {
        return name == 'Sangeeth';
    }
}
```

Then use like

```csharp
var result = "Sangeeth".IsAGoodBoy();
if(result)
{
	Console.WriteLine("Hey good boy");
}
```

<hr/>

## Proposed Improved Extension Support

How about introducing "PolyArg" extension methods, Where the extension method signature will match when we couple multiple arguments

- Here 'this' keyword can be used with multiple arguments making new signatures
- Could be generic as well

```csharp
public static class PolyArgExtensions
{
    public static bool IsAGoodBoyUnder24(this string name, this int val)
    {
        return name == 'Sangeeth' && val < 24;
    }
}
```

Then use like

- We could either use round brackets or any other syntax form to group them and call new poly-argument/multi-argument extension

```csharp
var result = ("Sangeeth", 24).IsAGoodBoyUnder24();
if(result)
{
	Console.WriteLine("Hey young good boy");
}
```

- Do you think this is a good idea?
- What could be potential implications?
- Will it conflict with any existing syntax matching?

Looking for a good discussion

**Comment by En3Tho** [comment]
This is basically an extension on ValueTuple.

**Comment by FaustVX** [comment]
You can already do that on extending tuple value `public static bool IsAGoodBoyUnder24(this (string name, int val) @this)`

[sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUBQG4TIAEyApgM6oA2CdAvHQBQBEAZQgwA5o0YIAFvzh1CAFgCUAOgCSzAIIBxAPa6MAIV0BPAKowMjZIt5KA3BSgAzXk1YclFAN4UAkPjEAJwCABKMJnQmuuiidKL6GHTApvwOFAC+FBT4AMx0gQBsBYR0AAq6bCaayKIAogAeCIwwzFC6rT4UdD0F+UXJ+mx0GjqJxuaW1rbSUMx8gQAMdDAQYIxysJw0bEoF+LPMXuS9dL4np734AOz7hyqr69w8QiLikjJ0AGRfd1JzKh2dAAPPIFI4Lj0suQMkA=)

**Comment by sangeethnandakumar** [comment]
Gotcha. Makes sense. Closing the discussion
